content/en/blog/2021/wasm-api-alpha/index.md

```diff
@@ -0,0 +1,120 @@
+  pluginConfig:
+    someSetting: true
+    someOtherSetting: false
+    youNameIt:
+    - first
+    - second
+  url: docker.io/your-org/your-filter:1.0.0
+{{< /text >}}
+
+There are a lot of similarities and a few differences between WasmPlugin and `EnvoyFilter`, so let’s go through the fields one by one. The above example deploys a Wasm module to all workloads (including gateway pods) that match it’s `selector` - this very much works the same as in an `EnvoyFilter`.
+
+The next field below that is the `phase`. This determines where in the proxy’s filter chain the Wasm module will be injected. We have defined four distinct phases for injection:
+
+* `AUTHN`: prior to any Istio authentication filters, meaning it is executed even before the remote mTLS certificates are verified.
+* `AUTHZ`: after the Istio authentication filters and before any authorization filters, i.e. before AuthorizationPolicies have been applied.
+* `STATS`: after all authorization filters and prior to the Istio stats filter.
+* `UNSPECIFIED_PHASE`: let the control plane decide where to insert. This will generally be at the end of the filter chain, right before the Router. This is the default value for this `phase` field.
+
+`pluginConfig` is used for configuring your Wasm plugin. Whatever you put into this field will be encoded in JSON and passed on to your filter, where you can access it in the configuration callback of the Proxy-Wasm SDKs. For example, you can retrieve the config on `onConfigure` in [C++ SDK](https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/fd0be8405db25de0264bdb78fae3a82668c03782/proxy_wasm_api.h#L329-L331), `on_configure` in [Rust SDK](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/v0.1.4/src/dispatcher.rs#L255) or `OnPluginStart` call back in [Go SDK](https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/v0.15.0/proxywasm/types/context.go#L74).
+
+The `url` field specifies where to pull the Wasm module. You’ll notice that the `url` in this case is a docker URI - this is because apart from loading Wasm modules via HTTP, HTTPS and the local file system (using file://), we are introducing the OCI image format as the preferred mechanism for distributing Wasm modules.
+
+## Wasm image specification
+
+We believe that containers are the ideal way to store, publish and manage proxy extensions, so we worked with Solo.io to extend their existing Proxy-Wasm container format with a variant that aims to be compatible with all registries and CLI toolchain. Depending on your processes, you can now either build your proxy extension containers using solo’s wasme tool or straight up with your existing container CLI tooling such as Docker CLI or [buildah](https://buildah.io/). Istio supports both variants.
+
+For detail, please refer to [the link here](https://github.com/solo-io/wasm/tree/master/spec), and learn [how to build OCI images](https://github.com/solo-io/wasm/tree/master/spec#how-can-i-build-images) that are consumable by Istio agent.
```

Comment:
I think we need to add some guide into istio-ecosystem repo and point there instead. That's the repo where we keep Istio supported guide.

Reply:
If you are okay, I would be happy to add doc like this https://github.com/mathetake/istio-wasm-api-demo to istio-ecosystem repo, put the link here. sg?

Reply:
Okay I'm on it. will raise in a hour or so.

Reply:
Thanks! I think it would be nice to just wrap one of the example (either the basic auth one or the plain example extension) about how to build it to an image. It can just be an part of existing guide: https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/write-a-wasm-extension-with-cpp.md#step-6-push-and-deploy-the-extension

Reply:
oh I am writing a separate document to put its link here like this:

```
# How to build Wasm OCI images

As of 1.12, Istio has supports for OCI images in its Wasm Plugin mechanism, and users are able to specify
any OCI registries where they store Wasm plugins as containers.

This document describes how to build OCI images which are consumable by Istio.

## Overview
```

so let's make the change to https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/write-a-wasm-extension-with-cpp.md#step-6-push-and-deploy-the-extension after we merge the dedicated document.

Reply:
Yep separated and focusing doc also sounds great to me. Thanks!

Reply:
https://github.com/istio-ecosystem/wasm-extensions/pull/74 a separated doc for this section

Reply:
done https://github.com/istio/istio.io/pull/10530/commits/9ec974537a286b58b9822209b40f4f35eff66da7